rts/lra/lra-coordinator/src/main/java/org/jboss/narayana/rts/lra/coordinator/domain/model/Transaction.java

```diff
@@ -0,0 +1,398 @@
+    private final URL parentId; // TODO save_state and restore_state
+    private final String clientId;
+    private List<LRARecord> pending;
+    private CompensatorStatus status; // reuse commpensator states for the LRA
+    private List<String> responseData;
+    private LocalTime cancelOn; // TODO make sure this acted upon during restore_state()
+    private ScheduledFuture<?> scheduledAbort;
+
+    public Transaction(String baseUrl, URL parentId, String clientId) throws MalformedURLException {
+        super(new Uid());
+
+        this.id = new URL(String.format("%s/%s", baseUrl, get_uid().fileStringForm()));
+        this.parentId = parentId;
+        this.clientId = clientId;
+        this.cancelOn = null;
+
+        status = null; // means the LRA is active
+
+        scheduler = Executors.newScheduledThreadPool(1);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Transaction)) return false;
+        if (!super.equals(o)) return false;
+
+        Transaction that = (Transaction) o;
+
+        return getId().equals(that.getId());
+    }
+
+    @Override
+    public int hashCode() {
+        int result = super.hashCode();
+        result = 31 * result + getId().hashCode();
+        return result;
+    }
+
+    public URL getId() {
+        return id;
+    }
+
+    public String getClientId() {
+        return clientId;
+    }
+
+    /**
+     * return the current status of the LRA
+     *
+     * @return null if the LRA is still active (not closing, cancelling, closed or cancelled
+     */
+    public CompensatorStatus getLRAStatus() {
+        return status;
+    }
+
+    public boolean isComplete() {
+        return status != null && status.equals(CompensatorStatus.Completed);
+    }
+
+    public boolean isCompensated() {
+        return status != null && status.equals(CompensatorStatus.Compensated);
+    }
+
+    public boolean isRecovering() {
+        return false;
+    } // TODO
+
+    private int closeLRA() {
+        return end(false);
+    }
+
+    private int cancelLRA() {
+        return end(true);
+    }
+
+    // in this version close need to run as blocking code {@link Vertx().executeBlocking}
+    public int end(/*Vertx vertx,*/ boolean compensate) {
+        int res = status();
+        boolean nested = !isTopLevel();
+
+        if (scheduledAbort != null) {
+            scheduledAbort.cancel(false);
+            scheduledAbort = null;
+        }
+
+        // nested compensators need to be remembered in case the enclosing LRA decides to compensate
+        // also save the list so that we can retrieve any response data after committing compensators
+        // if (nested)
+        savePendingList();
+
+        if ((res != ActionStatus.RUNNING) && (res != ActionStatus.ABORT_ONLY)) {
+            if (nested && compensate) {
+                /*
+                 * TODO this is wrong - we should be hooking into ActionType.NESTED ... but
```

Comment:
Is this still a problem?

Reply:
I don't think so - I will remove the comment